### PR TITLE
Add docs to stop processing hooks with async

### LINKED ISSUE #2075: Use symbol to denote early termination of request
## 🚀 Feature Proposal

Introduce `Symbol.for('fastify.end-request')` to terminate requests prior to reaching the defined request handler. When this symbol is received from a hook, or any function in the request life _prior_ to the defined request handler, Fastify should issue a `res.send()` that sends any headers and content currently attached to `res` to the requesting client.

This should also take the place of returning `undefined` to signal the same thing. Thus, when `undefined` is received from such a function, a deprecation notice/error/warning should be logged.

## Motivation

As highlighted in https://github.com/fastify/fastify/issues/2070 and https://github.com/fastify/help/issues/135 , it is unclear how one should terminate a request response before reaching the handler.

## Example

```js
fastify.addHook('onRequest', () => {
  // do something
  return Symbol.for('fastify.end-request')
})

// or

fastify.addHook('onRequest', (req, res, next) => {
  // do something
  next(null, Symbol.for('fastify.end-request'))
})

// or

fastify.addHook('onRequest', async (req, res) => {
  // do something
  return Symbol.for('fastify.end-request')
})
```

### PR DESCRIPTION
**Pull request by mcollina**
Fixes #2075

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Hooks.md [hunk 1]
```diff
@@ -232,6 +232,26 @@ fastify.addHook('onRequest', (request, reply, done) => {
 })
 ```
 
+If you are sending a response without `await` on it, make sure to always
```
Comment (jsumners) [inline docs/Hooks.md]: (GitHub won't let me add a comment to line 213)

On line 213 I think we should have more detail about _why_ these rules are in place and what is actually happening when they are followed. Even with the addition here on line 235 it's quite ambiguous as to what is going on.

Comment (mcollina) [reply docs/Hooks.md]: I've pushed some new text. Please check it out.

File: docs/Hooks.md [hunk 2]
```diff
@@ -210,7 +210,22 @@ fastify.addHook('preHandler', (request, reply, done) => {
...
 
 ### Respond to a request from a hook
-If needed, you can respond to a request before you reach the route handler, for example when implementing an authentication hook. If you are using `onRequest` or `preHandler` use `reply.send`; if you are using a middleware, use `res.end`.
+
+If needed, you can respond to a request before you reach the route handler,
+for example when implementing an authentication hook.
+Replying from an hook implies that the hook chain is __stopped__ and
+the rest of hooks and the handlers are not executed. If the hook is
+using the callback approach, i.e. it is not an `async` function or it
+return a `Promise`, it is as simple as calling `reply.send()` and avoid
```
Comment (jsumners) [inline docs/Hooks.md]:
```suggestion
returns a `Promise`, it is as simple as calling `reply.send()` and avoiding
```

File: docs/Hooks.md [hunk 3]
```diff
@@ -210,7 +210,22 @@ fastify.addHook('preHandler', (request, reply, done) => {
...
+calling the callback. If the hook is `async`, `reply.send()` __must__ be
+called _before_ the function returns/the promise resolves, otherwise the
```
Comment (jsumners) [inline docs/Hooks.md]:
```suggestion
called _before_ the function returns or the promise resolves, otherwise the
```

File: docs/Hooks.md [hunk 4]
```diff
@@ -210,7 +210,22 @@ fastify.addHook('preHandler', (request, reply, done) => {
...
+request will proceed; in case `reply.send()` is called outside of the
```
Comment (jsumners) [inline docs/Hooks.md]:
```suggestion
request will proceed. When `reply.send()` is called outside of the
```

File: docs/Hooks.md [hunk 5]
```diff
@@ -210,7 +210,22 @@ fastify.addHook('preHandler', (request, reply, done) => {
...
+promise chain, it is important to `return reply` otherwise the request .
```
Comment (jsumners) [inline docs/Hooks.md]: This line seems unfinished.

Comment (mcollina) [reply docs/Hooks.md]: Done, check it out.

File: test/hooks-async.js [hunk 6]
```diff
@@ -547,6 +547,25 @@ function asyncHookTest (t) {
 
     t.end()
   })
+
+  t.test('early termination, onRequest', async t => {
```
Comment (Eomm) [inline test/hooks-async.js]:
```suggestion
  t.test('early termination, onRequest async', async t => {
```

File: test/hooks.test.js [hunk 7]
```diff
@@ -2727,6 +2727,27 @@ test('onRegister hook should be called / 4', t => {
   })
 })
 
+test('early termination, onRequest', t => {
```
Comment (Eomm) [inline test/hooks.test.js]: Is a test for early termination with the signature `(request, reply, done)` already in place?

Comment (mcollina) [reply test/hooks.test.js]: Yes https://github.com/fastify/fastify/blob/c052c21db4b1d8b80eafd3435a6ba4ccdc9f97f8/test/hooks.test.js#L1213

**Comment by mcollina** [comment]
@Eomm are you ok?

**Pull request merged**